analysis/GreenSKU-Framework/src/carbon_model.py

```diff
@@ -0,0 +1,1003 @@
+
+    Returns:
+        The power consumption of the server.
+    """
+    if 'number' not in component:
+        component['number'] = 1
+    if not derate_curve:
+        derate_curve = lambda x: 1
+    return component['power'] * component['number'] * derate_curve(spec) * factor
+
+def get_dict_power(components: Dict[str, Any], derate_curve: Callable=None, 
+                   spec: int=100, factor: float=1.0, key_factors: Dict[str, float]={}) -> float:
+    """Calculate the total power consumption of a component.
+
+    Args:
+        components: A dictionary containing the components' power consumption and number.
+
+    Returns:
+        The total power consumption of the server.
+    """
+    total_power = 0
+    for key, component in components.items():
+        curr_factor = factor
+        if 'number' not in component:
+            component['number'] = 1
+        if key in key_factors:
+            curr_factor *= key_factors[key]
+        total_power += get_power_from_dict(component, 
+                                           derate_curve, 
+                                           spec,
+                                           factor=curr_factor)
+    return total_power
+
+def get_dict_capex(components: Dict[str, Any], cost='cost') -> float:
+    """Calculate the capex cost of a component.
+
+    Args:
+        components: A dictionary containing the components' cost and number.
+
+    Returns:
+        The capex cost of the server.
+    """
+    total_capex = 0
+    for key, component in components.items():
+        if 'number' not in component:
+            component['number'] = 1
+        total_capex += component[cost] * component['number']
+    return total_capex
+
+def get_list_capex(components: List[Dict[str, Any]]) -> float:
+    """Calculate the capex cost of a list of components.
+
+    Args:
+        components: A list of components.
+
+    Returns:
+        The capex cost of the server.
+    """
+    # Calculate the capex cost of the server
+    total_capex = 0
+    for component in components:
+        # if no number is specified, assume there is only one
+        if 'number' not in component:
+            component['number'] = 1
+        total_capex += component['cost'] * component['number']
+    return total_capex
+
+def index_cpu_data(vendor: str, cpu_type: str, core_count: int, data_source: str="data_sources") -> Dict[str, Any]:
+    """Index the CPU data for a given vendor, CPU type, and core count.
+    
+    Args:
+        vendor: The CPU vendor.
+        cpu_type: The CPU type.
+        core_count: The number of cores in the CPU.
+        data_source: The directory containing the CPU data.
+
+    Returns:
+        A dictionary containing the CPU data.
+    """
+    cpu_data = read_yaml(join_path(data_source, 'CPU.yaml'))
+    
+    data = None
+    spec_derates = None
+    for cpu_vendor in cpu_data:
```

Comment:
This is a nitpick but so much nesting is tough.
Either pass it into a DataFrame where you can search cleanly or do something like:
```
if cpu_vendor['vendor'] != vendor:
  continue

spec_derates = cpu_vendor['spec_derates']
for cpu_type_data in cpu_vendor['types']:
  if cpu_type_data['type'] != cpu_type:
    continue

```